ADAL/src/cache/ios/ADKeychainTokenCache.m

```diff
@@ -245,32 +245,98 @@ - (ADTokenCacheItem*)itemFromKeyhainAttributes:(NSDictionary*)attrs
 
 
 /*!
-    @param  item    The item to remove from the cache
+    @param  item    The item to be removed. Item with refresh token will be set as a tombstone, those without will be deleted.
     @param  error   (Optional) In the case of an error this will be filled with the
                     error details.
  
-    @return YES if the item was successfully removed, or was not in the cache. If NO
-            look
+    @return YES if the item was successfully tombstoned/deleted or not in the cache.
  */
 - (BOOL)removeItem:(nonnull ADTokenCacheItem *)item
              error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
 {
     RETURN_NO_ON_NIL_ARGUMENT(item);
     
+    //if item contains a refresh token, tombstone it;
+    //Otherwise delete it directly.
+    if (![NSString adIsStringNilOrBlank:item.refreshToken])
+    {
+        ADTokenCacheKey* key = [item extractKey:error];
+        if (!key)
+        {
+            return NO;
+        }
+        
+        // In layers above this a nil/blank user ID means we simply don't know who it is (thanks to ADFS)
+        // however for the purposes of adding users we still do need to have an account name, even if it
+        // is just blank.
+        NSString* userId = item.userInformation.userId;
+        if (!userId)
+        {
+            userId = @"";
+        }
+        
+        NSMutableDictionary* query = [self queryDictionaryForKey:key
+                                                          userId:userId
+                                                      additional:nil];
+        
+        //set the item to tombstone
+        ADTokenCacheItem* tombstoneItem = [item copy];
+        [self setCacheItemAsTombstone:tombstoneItem];
+        NSData* itemData = [NSKeyedArchiver archivedDataWithRootObject:tombstoneItem];
+        NSMutableDictionary * itemUpdate = [NSMutableDictionary dictionaryWithDictionary:@{ (id)kSecValueData : itemData,
+                                                                                              (id)kSecAttrAccessible : (id)kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly}];
+        [tombstoneItem release];
+        
+        //update the tombstone item to keychain
+        OSStatus status = SecItemUpdate((CFDictionaryRef)query, (CFDictionaryRef)itemUpdate);
+        return [ADKeychainTokenCache checkStatus:status details:@"Failed to remove item from keychain" error:error];
+    }
+    else
```

Comment:
You can put this up at the top as a if ([NSString adIsStringNilOrBlank:item.refreshToken]) block and early return, then the rest of the code doesn't need to be indented!